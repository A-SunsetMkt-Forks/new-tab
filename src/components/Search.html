<!--
  XXX: The h2 titles are implemented abnormally to get around a current Svelte
  limitation where it creates multiple text nodes instead of realising there's
  multiple text nodes that could be combined into one. Once Svelte has this
  optimisation these titles should be refactored.
-->

<div class="container">
  <input
    id="search"
    bind:value="_searchText"
    ref:s
    placeholder="Search browser…"
    autocomplete="off"
    on:keyup="_cancelSearch(event.key)"
  >

  {#each _order as _section}
    {#if _section === 'Open Tabs'}

      <!-- Open Tabs -->

      <h2>{`Open Tabs (${_tabs.length}/${_tabsRaw.length})`}</h2>

      {#each _tabs as _item}
        <!-- XXX: This is the same as <LinkItem> but with a click handler -->
        <a href="{_item.url}" title="{_item.title}" on:click="_onTabClick(_item, event)">
          <img src="chrome://favicon/{_item.url}">
          {_item.title}
        </a>
      {/each}
    {:elseif _section === 'Top Sites'}

      <!-- Top Sites -->

      <h2>{`Top Sites (${_topSites.length}/${_topSitesRaw.length})`}</h2>

      {#each _topSites as _item}
        <LinkItem _url="{_item.url}" _title="{_item.title}" />
      {/each}
    {:elseif _isSearching && _section === 'Bookmarks'}

      <!-- Bookmarks -->

      <h2>{`Bookmarks (${_bookmarks.length}/${_bookmarksRaw.length})`}</h2>

      {#each _bookmarks as _item}
        <LinkItem _url="{_item.url}" _title="{_item.title}" />
      {/each}

      {#if _bookmarks.length < _bookmarksRaw.length}
        <button on:click="_loadMoreBookmarks(_bookmarks.length)">Show more ▾</button>
      {/if}
    {:elseif _isSearching && _section === 'History'}

      <!-- History -->

      <h2>{`History (${_history.length}/${_historyRaw.length})`}</h2>

      {#each _history as _item}
        <LinkItem _url="{_item.url}" _title="{_item.title}" />
      {/each}

      {#if _history.length < _historyRaw.length}
        <button on:click="_loadMoreHistory(_history.length)">Show more ▾</button>
      {/if}
    {/if}
  {/each}
</div>

<script>
  import { DEFAULT_ORDER, debounce } from '../common.js';
  import LinkItem from './LinkItem.html';

  const SEARCH_DEBOUNCE_DELAY = 260; // ms
  const DEFAULT_RESULTS_AMOUNT = 10;
  const MORE_RESULTS_AMOUNT = 100;

  // URL protocol regex
  const protocol = /^.*?:\/\//;

  /**
   * Check for a matching title or URL.
   * @this {string} Search query
   * @param {{ title: string, url: string }} item
   * @returns {boolean}
   */
  function searchFilter({ title, url }) {
    return title.toLowerCase().indexOf(this) > -1
      // remove URL protocol since it's unlikely to be searched for
      || url.replace(protocol, '').toLowerCase().indexOf(this) > -1;
  }

  export default {
    components: {
      LinkItem,
    },
    data: () => ({
      _order: DEFAULT_ORDER,
      _searchText: '',
      _bookmarksRaw: [],
      _bookmarks: [],
      _historyRaw: [],
      _history: [],
      _tabsRaw: [],
      _tabs: [],
      _topSites: [],
      _topSitesRaw: [],
    }),
    computed: {
      _isSearching: ({ _searchText }) => !!_searchText,
    },

    oncreate() {
      chrome.storage.local.get(null, (settings) => {
        /* eslint-disable dot-notation */ // prevent closure mangling
        if (settings['o']) {
          this.set({ _order: settings['o'] });
        }
        /* eslint-enable */
      });

      const getTabs = () => this._getTabs();

      getTabs();

      chrome.topSites.get((sites) => {
        this.set({
          _topSites: sites,
          _topSitesRaw: sites,
        });
      });

      this._onSearch = debounce(this._search, SEARCH_DEBOUNCE_DELAY);

      // bring focus to search input when clicking anywhere on the page
      window.addEventListener('click', () => {
        this.refs.s.focus();
      });

      // update tab list on tab events
      chrome.tabs.onUpdated.addListener(getTabs);
      chrome.tabs.onRemoved.addListener(getTabs);
      chrome.tabs.onMoved.addListener(getTabs);
      chrome.tabs.onAttached.addListener(getTabs);
    },

    // FIXME: When search is active, rerun when tab events trigger and _getTabs() is run
    onstate({ changed, previous }) {
      if (!previous) return;

      if (changed._searchText) {
        this._onSearch();
      }
    },

    methods: {
      _search() { // eslint-disable-line consistent-return
        const { _searchText, _tabsRaw, _topSitesRaw } = this.get();

        // reset search when input is empty
        if (_searchText === '') {
          return this.set({
            _bookmarks: [],
            _history: [],
            _tabs: _tabsRaw,
            _topSites: _topSitesRaw,
          });
        }

        // search history (run first since it takes the longest)
        chrome.history.search({ text: _searchText }, (results) => {
          this.set({
            _historyRaw: results,
            _history: results.slice(0, DEFAULT_RESULTS_AMOUNT),
          });
        });

        // search bookmarks
        chrome.bookmarks.search(_searchText, (results) => {
          this.set({
            _bookmarksRaw: results,
            _bookmarks: results.slice(0, DEFAULT_RESULTS_AMOUNT),
          });
        });

        this.set({
          // search open tabs
          _tabs: _tabsRaw.filter(searchFilter, _searchText),
          // search top sites
          _topSites: _topSitesRaw.filter(searchFilter, _searchText),
        });
      },

      /**
       * Query the browser for a list of the currently open tabs.
       */
      _getTabs() {
        chrome.tabs.query({}, (tabs) => {
          if (this.get()._searchText === '') {
            this.set({
              _tabs: tabs,
              _tabsRaw: tabs,
            });
          } else {
            this.set({ _tabsRaw: tabs });
          }
        });
      },

      _onTabClick(tab, event) {
        event.preventDefault();

        // update current tab
        chrome.tabs.update(tab.id, { active: true });

        // switch active window if the tab isn't in the current window
        chrome.windows.getCurrent({}, (currentWindow) => {
          if (currentWindow.id !== tab.windowId) {
            chrome.windows.update(tab.windowId, { focused: true });
          }
        });

        // close this NTP
        chrome.tabs.getCurrent((currentTab) => {
          chrome.tabs.remove(currentTab.id);
        });
      },

      _cancelSearch(key) {
        if (key === 'Escape') {
          const { _tabsRaw, _topSitesRaw } = this.get();
          this.set({
            _searchText: '',
            _bookmarks: [],
            _history: [],
            _tabs: _tabsRaw,
            _topSites: _topSitesRaw,
          });
        }
      },

      _loadMoreBookmarks(length) {
        this.set({ _bookmarks: this.get()._bookmarksRaw.slice(0, length + MORE_RESULTS_AMOUNT) });
      },

      _loadMoreHistory(length) {
        this.set({ _history: this.get()._historyRaw.slice(0, length + MORE_RESULTS_AMOUNT) });
      },
    },
  };
</script>

<style type="text/postcss" global>
  .container {
    max-width: 800px;
    padding: 0 18px;
    margin: 0 auto;
  }

  button,
  #search {
    box-sizing: border-box;
    width: 100%;
    padding: 11px 20px;
    margin: 0 0 18px;
    font-size: 22px;
    color: var(--t);
    background: var(--c0);
    border: 0;
    box-shadow: var(--s);
  }

  /* search input placeholder */
  ::placeholder {
    color: var(--c2);
  }

  /* load more buttons */
  button {
    width: initial;
    margin-top: 9px;
    cursor: pointer;
  }
</style>
