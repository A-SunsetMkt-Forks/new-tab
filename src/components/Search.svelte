<script>
  import { DEFAULT_ORDER, debounce } from '../utils';
  import SearchResults from './SearchResults.svelte';

  const SEARCH_DEBOUNCE_DELAY_MS = 260;
  const DEFAULT_RESULTS_AMOUNT = 10;

  /** URL protocol regex. */
  const protocolRe = /^.*?:\/\//;

  /** Sorted bookmarks list. */
  let bList = [];
  /** Raw bookmarks list. */
  let bRaw = [];
  /** Sorted history list. */
  let hList = [];
  /** Raw history list. */
  let hRaw = [];
  /** Sorted top sites list. */
  let tList = [];
  /** Raw top sites list. */
  let tRaw = [];
  let tabsList = [];
  let tabsRaw = [];
  let order = [];
  let searchText = '';

  /** Query the browser for a list of the currently open tabs. */
  const getTabs = () => {
    chrome.tabs.query({}, (tabs) => {
      if (searchText === '') {
        // eslint-disable-next-line no-multi-assign
        tabsList = tabsRaw = tabs;
      } else {
        tabsRaw = tabs;
      }
    });
  };

  /** @param {MouseEvent} obj - A link item mouse click event. */
  const handleTabClick = ({ target }) => {
    const windowId = target.getAttribute('w');

    // Update current tab
    chrome.tabs.update(+target.id, { active: true });

    // Switch active window if the tab isn't in the current window
    chrome.windows.getCurrent({}, (currentWindow) => {
      if (currentWindow.id !== windowId) {
        chrome.windows.update(windowId, { focused: true });
      }
    });

    // Close this NTP
    chrome.tabs.getCurrent((currentTab) => {
      chrome.tabs.remove(currentTab.id);
    });
  };

  const maybeCancelSearch = (event) => {
    if (event.key === 'Escape') {
      searchText = '';
      // eslint-disable-next-line no-multi-assign
      bList = hList = [];
      tabsList = tabsRaw;
      tList = tRaw;
    }
  };

  /**
   * Check if either title or URL match a query.
   *
   * @this {string} Search query.
   * @param {object} obj - Options.
   * @param {string} obj.title - Title to match against.
   * @param {string} obj.url - URL to match against.
   * @returns {boolean} True if a match is found.
   */
  const searchFilter = ({ title, url }) =>
    title.toLowerCase().indexOf(this) > -1
    // Remove URL protocol since it's unlikely to be searched for
    || url.replace(protocolRe, '').toLowerCase().indexOf(this) > -1;


  const doSearch = () => {
    // Reset search when input is empty
    if (!searchText) {
      // eslint-disable-next-line no-multi-assign
      bList = hList = [];
      tabsList = tabsRaw;
      tList = tRaw;
      return;
    }

    // Search history (run first since it takes the longest)
    chrome.history.search({ text: searchText }, (results) => {
      hRaw = results;
      hList = results.slice(0, DEFAULT_RESULTS_AMOUNT);
    });

    // Search bookmarks
    chrome.bookmarks.search(searchText, (results) => {
      bRaw = results;
      bList = results.slice(0, DEFAULT_RESULTS_AMOUNT);
    });

    // Search open tabs
    tabsList = tabsRaw.filter(searchFilter, searchText);
    // Search top sites
    tList = tRaw.filter(searchFilter, searchText);
  };

  const onSearch = debounce(doSearch, SEARCH_DEBOUNCE_DELAY_MS);

  chrome.storage.local.get(null, (settings) => {
    order = settings['o'] || DEFAULT_ORDER;
  });

  getTabs();

  chrome.topSites.get((sites) => {
    // eslint-disable-next-line no-multi-assign
    tList = tRaw = sites;
  });

  // Update tab list on tab events
  chrome.tabs.onUpdated.addListener(getTabs);
  chrome.tabs.onRemoved.addListener(getTabs);
  chrome.tabs.onMoved.addListener(getTabs);

  $: if (searchText) onSearch();
</script>

<style type="text/postcss">
  /* "Load more" buttons */
  :global(button) {
    width: initial;
    margin-top: 9px;
    cursor: pointer;
  }

  :global(.container) {
    max-width: 800px;
    margin: 0 auto;
    padding: 0 18px;
  }

  /* stylelint-disable no-descending-specificity */
  :global(button),
  :global(#search) {
    box-sizing: border-box;
    width: 100%;
    margin: 0 0 35px;
    padding: 11px 20px;
    color: var(--t);
    font-size: 21px;
    background: var(--c2);
    border: 0;
    border-radius: 24px;
    outline: none;

    &:hover,
    &:focus {
      background: var(--c1);
    }
  }
  /* stylelint-enable */
</style>

<div class="container">
  <input
    id="search"
    bind:value="{searchText}"
    placeholder="Search browserâ€¦"
    autocomplete="off"
    on:keyup="{maybeCancelSearch}"
  >

  {#each order as resultSection}
    {#if resultSection === 'Open Tabs'}
      <h2>{`Open Tabs (${tabsList.length}/${tabsRaw.length})`}</h2>

      <div on:click|capture="{handleTabClick}">
        {#each tabsList as _node}
          <a
            href="{_node.url}"
            id="{_node.id}"
            w="{_node.windowId}"
            title="{_node.title}"
          >
            <img src="chrome://favicon/{_node.url}" />
            {_node.title}
          </a>
        {/each}
      </div>
    {:else if resultSection === 'Top Sites'}
      <SearchResults name="Top Sites" list="{tList}" raw="{tRaw}" />
    {:else if !!searchText && resultSection === 'Bookmarks'}
      <SearchResults name="Bookmarks" list="{bList}" raw="{bRaw}" />
    {:else if !!searchText && resultSection === 'History'}
      <SearchResults name="History" list="{hList}" raw="{hRaw}" />
    {/if}
  {/each}
</div>
