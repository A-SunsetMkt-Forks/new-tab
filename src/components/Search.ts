// TODO: Implement search and list filtering

// TODO: Implement search reset

import h from 'stage0';
import { SearchResult } from './SearchResult';
import { debounce, DEFAULT_ORDER } from '../utils';

function searchLists(text) {
  console.log('## SEARCH TEXT', text);
}

const doSearch = debounce(searchLists);

const view = h`
  <div class=container>
    <input id=search placeholder="Search browser..." autocomplete=off #input>
    <svg xmlns=http://www.w3.org/2000/svg width=24 height=24 viewBox="0 0 24 24" class=s-icon>
      <circle cx=11 cy=11 r=8 />
      <line x1=24 y1=24 x2=16.65 y2=16.65 />
    </svg>
  </div>
`;

export function Search() {
  const root = view;
  const { input } = view.collect(root);

  const state = {};

  input.oninput = (event) => {
    doSearch(input.value);
  };

  input.onkeyup = (event) => {
    if (event.key === 'Escape') {
      // reset filter
    }
  };

  const update = () => {
    chrome.tabs.query({}, (tabs) => {
      state['Open Tabs'].update(tabs);
    });
  };

  // Get user settings
  chrome.storage.local.get(null, (settings) => {
    const order = settings.o || DEFAULT_ORDER;

    order.forEach((name) => {
      state[name] = root.appendChild(SearchResult(name, []));
    });

    update();

    if (state['Top Sites']) {
      chrome.topSites.get((sites) => {
        state['Top Sites'].update(sites);
      });
    }
  });

  // Update tab list on tab events
  chrome.tabs.onUpdated.addListener(update);
  chrome.tabs.onRemoved.addListener(update);
  chrome.tabs.onMoved.addListener(update);

  return root;
}
