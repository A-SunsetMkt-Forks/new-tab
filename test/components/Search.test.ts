import Search from '../../src/components/Search.svelte';

describe('Search component', () => {
  it('renders correctly', () => {
    const target = document.createElement('div');
    new Search({ target });
    expect(target.innerHTML).toMatchSnapshot();
  });
});
