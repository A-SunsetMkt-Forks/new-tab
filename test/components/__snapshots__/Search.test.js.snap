// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component renders correctly 1`] = `
"<div class=\\"c\\"><input id=\\"s\\" placeholder=\\"Search browser...\\" autocomplete=\\"off\\">

  
  <p class=\\"g\\">No matching Open Tabs</p>

  
  

  
  

  
  </div>"
`;
